igvc_navigation/src/path_follower/smooth_control.cpp

```diff
@@ -0,0 +1,252 @@
+
+    // Calculate control using Control Law
+    Action action = getAction(angles.delta, angles.theta, state, simulation_target);
+    action.dt = dt.toSec();
+
+    if (i == 0)
+    {
+      target = simulation_target;
+      getWheelVelocities(vel, action);
+    }
+
+    propogateState(action, state);
+
+    geometry_msgs::PoseStamped pose;
+    pose.pose.position.x = state.x;
+    pose.pose.position.y = state.y;
+    pose.pose.orientation = state.quat();
+    tf::Quaternion quat;
+    tf::quaternionMsgToTF(pose.pose.orientation, quat);
+    double r, p, y;
+    tf::Matrix3x3(quat).getRPY(r, p, y);
+    pose.header.stamp = time;
+    trajectory.poses.emplace_back(pose);
+
+    time += dt;
+  }
+}
+
+void SmoothControl::propogateState(const Action& action, RobotState& state)
+{
+  double v = action.v;
+  double w = action.w;
+  double dt = action.dt;
+
+  // pose after applying action to current pose
+  Eigen::Vector3d resultant_pose;
+
+  if (std::abs(w) > 1e-10)
+  {
+    // calculate instantaneous center of curvature (ICC = [ICCx, ICCy])
+    double R = v / w;
+    double ICCx = state.x - (R * sin(state.yaw));
+    double ICCy = state.y + (R * cos(state.yaw));
+
+    using namespace Eigen;
+    Matrix3d T;
+    double wdt = w * dt;
+    T << cos(wdt), -sin(wdt), 0, sin(wdt), cos(wdt), 0, 0, 0, 1;
+    Vector3d a(state.x - ICCx, state.y - ICCy, state.yaw);
+    Vector3d b = T * a;
+    Vector3d c = b + Vector3d(ICCx, ICCy, wdt);
+    igvc::fit_to_polar(c[2]);
+
+    resultant_pose << c[0], c[1], c[2];
+  }
+  else
+  {
+    resultant_pose << state.x + (cos(state.yaw * v * dt)), state.y + (sin(state.yaw) * v * dt), state.yaw;
+  }
+
+  state.setState(resultant_pose);
+}
+
+Action SmoothControl::getAction(double delta, double theta, const RobotState& state, const RobotState& target)
+{
+  // get egocentric polar coordinates for delta and theta
+
+  // adjust both angles to lie between -PI and PI
+  igvc::fit_to_polar(delta);
+  igvc::fit_to_polar(theta);
+
+  // calculate the radius of curvature, K
+  double d = state.distTo(target);  // euclidian distance to target
+  double K = k2_ * (delta - atan(-k1_ * theta));
+  K += (1 + (k1_ / (1 + pow(k1_ * theta, 2)))) * sin(delta);
+  K /= -d;
+
+  // calculate angular velocity using radius of curvature and target velocity
+  double w = K * target_velocity_;
+  return Action{ target_velocity_, w, 0 };
+}
+
+Egocentric_angles SmoothControl::getEgocentricAngles(const nav_msgs::PathConstPtr& path, unsigned int path_index,
+                                                     const Eigen::Vector3d& los, const Eigen::Vector3d& heading)
+{
+  // get egocentric polar angle of los relative to heading
+  double delta;
+  igvc::compute_angle(delta, los, heading);
+
+  // get i and j components of target orientation vector (res_orientation)
+  double distance = 0;
+  geometry_msgs::Point point1, point2;
+  while (path_index < path->poses.size() - 1)
```

Comment:
This logic is duplicated in the getTargetPosition function.

Reply:
Removed getEgocentricAngles to use `atan2` instead.